pumpkin-inventory/src/lib.rs

```diff
@@ -206,32 +215,51 @@ impl<'a> Container for OptionallyCombinedContainer<'a, 'a> {
     }
 
     fn window_name(&self) -> &'static str {
-        self.container
-            .as_ref()
-            .map(|container| container.window_name())
-            .unwrap_or(self.inventory.window_name())
-    }
-
-    fn all_slots(&mut self) -> Vec<&mut Option<ItemStack>> {
-        let slots = match &mut self.container {
-            Some(container) => {
-                let mut slots = container.all_slots();
-                slots.extend(self.inventory.all_combinable_slots_mut());
-                slots
-            }
-            None => self.inventory.all_slots(),
+        let Some(container) = &self.container else {
+            return self.inventory.window_name();
         };
-        slots
+        container.window_name()
     }
 
-    fn all_slots_ref(&self) -> Vec<Option<&ItemStack>> {
+    fn iter_slots_mut<'s>(&'s mut self) -> Box<(dyn Iterator<Item = &mut Option<ItemStack>> + 's)> {
+        match &mut self.container {
+            Some(container) => Box::new(
+                container
+                    .iter_slots_mut()
+                    .chain(self.inventory.iter_slots_mut()),
```

Comment:
If container is `Some` on `OptionallyCombinedContainer` only chain it with the main inventory slots on `PlayerInventory`. The client is not able to change things (like armour) while inside of a container. 

See https://wiki.vg/Inventory for a nice visualisation of where each slot goes, depending on which container that is opened.